File: rules-interpreter/src/main/java/io/github/iamrenny/ruleflow/evaluators/DateDiffContextEvaluator.java
package io.github.iamrenny.ruleflow.evaluators;

import io.github.iamrenny.ruleflow.RuleFlowLanguageParser;
import io.github.iamrenny.ruleflow.RuleFlowLanguageParser.DateDiffContext;
import io.github.iamrenny.ruleflow.visitors.Visitor;

import java.time.Duration;
import java.time.LocalDateTime;
import java.time.ZonedDateTime;
import java.time.temporal.ChronoUnit;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class DateDiffContextEvaluator implements ContextEvaluator<RuleFlowLanguageParser.DateDiffContext> {
    private static final Logger logger = LoggerFactory.getLogger(DateDiffContextEvaluator.class);
    @Override
    public Object evaluate(DateDiffContext ctx, Visitor visitor) {
        ZonedDateTime left = (ZonedDateTime) visitor.visit(ctx.left);
        ZonedDateTime right = (ZonedDateTime) visitor.visit(ctx.right);

        if (left == null || right == null) {
            logger.debug("DateDiff: left={}, right={}", left, right);
            return null;
        }

        long result;

        if (ctx.MINUTE() != null) {
            result = Duration.between(left, right).toMinutes();
        } else if (ctx.HOUR() != null) {
            result = Duration.between(left, right).toHours();
        } else if (ctx.DAY() != null) {
            result = Duration.between(left, right.truncatedTo(ChronoUnit.DAYS)).toDays();
        } else {
            throw new RuntimeException("Interval not supported in " + ctx.getText());
        }

        long abs = Math.abs(result);

        logger.debug("DateDiff: left={}, right={}, result={}", left, right, abs);
        return abs;
    }
}

File: rules-interpreter/src/main/kotlin/com/rappi/fraud/rules/parser/evaluators/DateDiffContextEvaluator.kt
package com.rappi.fraud.rules.parser.evaluators

import com.rappi.fraud.analang.ANAParser.DateDiffContext
import com.rappi.fraud.rules.parser.visitors.Visitor
import java.time.Duration
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import kotlin.math.absoluteValue

class DateDiffContextEvaluator : ContextEvaluator<DateDiffContext> {

    override fun evaluate(ctx: DateDiffContext, visitor: Visitor): Any {
        val valLeft = visitor.visit(ctx.left)
        val valRight = visitor.visit(ctx.right)
        val left =  LocalDateTime.parse(valLeft.toString())
        val right = LocalDateTime.parse(valRight.toString())

        return when {
                    ctx.MINUTE() != null ->  Duration.between(left, right).toMinutes()
                    ctx.HOUR() != null -> Duration.between(left, right).toHours()
                    ctx.DAY() != null -> Duration.between(left, right.truncatedTo(ChronoUnit.DAYS)).toDays()
                    else -> throw RuntimeException("Interval not supoorted in ${ctx.text}")
                }.absoluteValue

    }

}